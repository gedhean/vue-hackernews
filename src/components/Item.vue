<template>
  <li>
    <a class="title" :href="item.url" target="_blank">
      {{ item.title }}
    </a>
    <p class="news-meta">
      <span class="type">{{ item.type }}</span>
      by
      <span class="author"> {{ item.by }}</span>
      - with score
      <span class="score">{{ item.score || 0 }}</span>
    </p>
  </li>
</template>

<script>
export default {
  props: {
    item: Object
  }
};
</script>
<style lang="scss" scoped>
.news-meta {
  font-size: 16px;

  .type {
    text-transform: capitalize;
  }

  .author {
    text-transform: lowercase;
    font-style: italic;
  }

  .score {
    font-weight: bold;
  }
}
</style>
