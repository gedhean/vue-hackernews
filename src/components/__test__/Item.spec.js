import Item from "./../Item.vue";
import { shallowMount } from "@vue/test-utils";

const item = {
  title: "news-title",
  by: "jedi",
  url: "http://www.example.com/news",
  score: 10
};

describe("Item.vue", () => {
  test("render prop item.title", () => {
    const wrapper = shallowMount(Item, { propsData: { item } });

    expect(wrapper.text()).toContain(item.title);
  });

  test("render prop item.by", () => {
    const wrapper = shallowMount(Item, { propsData: { item } });

    expect(wrapper.text()).toContain(item.by);
  });

  test("render prop item.score", () => {
    const wrapper = shallowMount(Item, { propsData: { item } });

    expect(wrapper.text()).toContain(item.score);
  });

  test("render a link to prop item.url with text item.title", () => {
    const wrapper = shallowMount(Item, { propsData: { item } });

    const link = wrapper.find("a.title");

    expect(link.text()).toContain(item.title);
    expect(link.attributes().href).toBe(item.url);
  });

  test("render time since item was posted", () => {
    const dateNow = jest.spyOn(Date, "now");
    const dateNowTime = new Date("2020");
    dateNow.mockImplementation(() => dateNowTime);

    const TEN_MINUTES_IN_SECONDS = 10 * 60;
    const time = Date.now() / 1000 - TEN_MINUTES_IN_SECONDS;
    const item = { time };
    const wrapper = shallowMount(Item, { propsData: { item } });

    expect(wrapper.text()).toContain("10 minutes ago");
    dateNow.mockRestore();
  });

  test("render news domain", () => {
    const wrapper = shallowMount(Item, {
      propsData: { item: { url: "http://www.example.com/news" } }
    });

    expect(wrapper.text()).toContain("example.com");
  });
});
