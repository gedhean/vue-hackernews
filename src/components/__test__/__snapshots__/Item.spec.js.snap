// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item.vue render correctly 1`] = `
<li><a href="https://gedhean.dev/tdd" target="_blank" class="title">
    learning tdd with vue.js
  </a>
  <p class="news-meta">
    by
    <span class="author"> jedi</span> <span class="time"> 2 minutes ago</span> <span class="domain"> (gedhean.dev)</span>
    - score
    <span class="score">1000</span></p>
</li>
`;
